Add VIP mu-plugins and object-cache installation

srtfisher opened the pull request:
Aims to complement the work in https://github.com/alleyinteractive/deploy/pull/4 to generate a shared bash script that set up a site for testing. Also fills in the gaps for non-Mantle testing framework-based solutions to use the same script.

## Changes

- Option to install https://github.com/Automattic/vip-go-mu-plugins-built.git from the script.
- Option to install `object-cache.php` from the script.
- Option to install core's testing suite (https://develop.svn.wordpress.org) to `WP_TESTS_DIR`. This is disabled by default.
- Adds documentation for all the arguments and available environmental variables that can be set by the user.

The next step here will be to add support to Mantle Testkit to control these options and allow projects to install `mu-plugins` and `object-cache.php` all within their `tests/bootstrap.php`.

srtfisher committed:
Adding shellcheck

srtfisher committed:
Adding basis for installing VIP mu plugins and object-cache

srtfisher committed:
Add testing

srtfisher committed:
Wire up the core test suite

srtfisher committed:
Comment

srtfisher committed:
Merging in develop

srtfisher committed:
Clarify doc

srtfisher committed:
README

srtfisher committed:
Testing CI

srtfisher committed:
Ignore if not able to delete

srtfisher committed:
Remove rm

srtfisher committed:
Allow error?

srtfisher committed:
Once more

srtfisher committed:
Debug current path

srtfisher committed:
Debug current path

srtfisher committed:
Adjusting test script

srtfisher committed:
Setting WP_CORE_DIR

srtfisher committed:
Revert back to classic object-cache

srtfisher committed:
Formatting and addt debug

benpbolton commented on install-wp-tests.sh:
```diff
@@ -42,11 +152,9 @@ download() {
-if [[ $WP_VERSION =~ ^[0-9]+\.[0-9]+\-(beta|RC)[0-9]+$ ]]; then
-	WP_BRANCH=${WP_VERSION%\-*}
-	WP_TESTS_TAG="branches/$WP_BRANCH"
+set +e
```
I never like seeing the removal of error exits without understanding why... can you elaborate on why we need to no longer care about error exit codes here?

benpbolton commented on install-wp-tests.sh:
```diff
@@ -72,20 +180,22 @@ fi
-		download https://wordpress.org/nightly-builds/wordpress-latest.zip  $CACHEDIR/wordpress-nightly/wordpress-nightly.zip
-		unzip -q $CACHEDIR/wordpress-nightly/wordpress-nightly.zip -d $CACHEDIR/wordpress-nightly/
-		mv $CACHEDIR/wordpress-nightly/wordpress/* $WP_CORE_DIR
+		mkdir -p "$CACHEDIR"/wordpress-nightly
```
```suggestion
		mkdir -p "$CACHEDIR/wordpress-nightly"
```
just reads a bit cleaner to wrap the entire string here

benpbolton commented on install-wp-tests.sh:
```diff
@@ -72,20 +180,22 @@ fi
+		mkdir -p "$CACHEDIR"/wordpress-nightly
+		download "https://wordpress.org/nightly-builds/wordpress-latest.zip"  "$CACHEDIR/wordpress-nightly/wordpress-nightly.zip"
+		unzip -q "$CACHEDIR/wordpress-nightly/wordpress-nightly.zip" -d "$CACHEDIR/wordpress-nightly/"
+		mv "$CACHEDIR"/wordpress-nightly/wordpress/* $WP_CORE_DIR
```
```suggestion
		mv "$CACHEDIR/wordpress-nightly/wordpress/"* "$WP_CORE_DIR"
```
this is an odd-duck... if you inclose the glob, you won't get what you want, so leaving the double quotes around the variable may be just fine... but we should probably enclose all variables in double quotes if we can...

benpbolton commented on install-wp-tests.sh:
```diff
@@ -135,12 +247,50 @@ install_config() {
+		# set up testing suite
+		mkdir -p $WP_TESTS_DIR
+		svn co --quiet https://develop.svn.wordpress.org/${WP_TESTS_TAG}/tests/phpunit/includes/ $WP_TESTS_DIR/includes
+		svn co --quiet https://develop.svn.wordpress.org/${WP_TESTS_TAG}/tests/phpunit/data/ $WP_TESTS_DIR/data
```
Two different variable reference styles here, they both work, but we should surround with quotes, regardless of approach

benpbolton commented on install-wp-tests.sh:
```diff
@@ -135,12 +247,50 @@ install_config() {
+		sed $ioption "s/youremptytestdbnamehere/$DB_NAME/" "$WP_TESTS_DIR"/wp-tests-config.php
+		sed $ioption "s/yourusernamehere/$DB_USER/" "$WP_TESTS_DIR"/wp-tests-config.php
+		sed $ioption "s/yourpasswordhere/$DB_PASS/" "$WP_TESTS_DIR"/wp-tests-config.php
+		sed $ioption "s|localhost|${DB_HOST}|" "$WP_TESTS_DIR"/wp-tests-config.php
```
```suggestion
		sed ${ioption "s:dirname( __FILE__ ) . '/src/':'$WP_CORE_DIR/':" "$WP_TESTS_DIR"/wp-tests-config.php
		sed ${ioption} "s/youremptytestdbnamehere/$DB_NAME/" "$WP_TESTS_DIR"/wp-tests-config.php
		sed ${ioption} "s/yourusernamehere/$DB_USER/" "$WP_TESTS_DIR"/wp-tests-config.php
		sed ${ioption} "s/yourpasswordhere/$DB_PASS/" "$WP_TESTS_DIR"/wp-tests-config.php
		sed ${ioption} "s|localhost|${DB_HOST}|" "$WP_TESTS_DIR"/wp-tests-config.php
```

just drawing more visual attention to the variable switches...

benpbolton commented on install-wp-tests.sh:
```diff
@@ -166,6 +316,55 @@ install_db() {
+			--depth=1 \
+			https://github.com/Automattic/vip-go-mu-plugins-built.git mu-plugins
+	else
+		yellow "VIP Go mu-plugins already exists, attempting to update"
```
🤔  there is a scenario here where someone is using mantle, but they've got a submoduled' or funky mu-plugins source that is *NOT* VIP... perhaps we could test whether the git remote was Automattic's?

benpbolton commented on install-wp-tests.sh:
```diff
@@ -166,6 +316,55 @@ install_db() {
-install_db
\ No newline at end of file
+install_db
+install_mu_plugins
```
I almost prefer an explicit mention of VIP here, eg. like `install_vip_mu_plugins` or `install_mu_plugins_vip` or something similar... to clarify that other use-cases and hosts may serve mu_plugins differently, but the purpose of these is to install vip go's mu-plugins?

benpbolton commented on README.md:
```diff
@@ -2,4 +2,9 @@
 Contains the continuous integration scripts used inside the Mantle Framework and
 Mantle packages. Documentation can be found inside the [Mantle
-documentation](https://mantle.alley.co/).
+documentation](https://mantle.alley.com/).
```
good update:)

benpbolton commented on bin/test-script.sh:
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+set -e
```
I don't mind these up in line 1 following the shebang #!/bin/bash -e

benpbolton commented on install-wp-tests.sh:
```diff
@@ -1,37 +1,147 @@
-CACHEDIR=$(echo $CACHEDIR | sed -e "s/\/$//")
+function red {
+	# red text to stderr
+	echo "$@" | sed $'s,.*,\e[31m&\e[m,' | >&2 xargs -0 printf
```
i like seeing these output to stderr 👍

benpbolton commented on install-wp-tests.sh:
```diff
@@ -72,20 +180,22 @@ fi
-	if [ -d $WP_CORE_DIR ]; then
-		if [ -f $WP_CORE_DIR/wp-load.php ]; then
+	if [ -d "$WP_CORE_DIR" ]; then
+		if [ -f "$WP_CORE_DIR"/wp-load.php ]; then
```
```suggestion
		if [ -f "$WP_CORE_DIR/wp-load.php" ]; then
```

just preferring the full wrap here

benpbolton reviewed (approved):
Great stuff! There was a lot to review (so I might've missed stuff). syntactically, I want to see quotes used like template strings (variable inside the full quoted line where possible). logically, I want to pay special attention to where we set `-e` and `+e` and understand why... philosophically, I wonder if it makes sense to allow for scenarios where `mu-plugins` are in-use outside of VIP's environment (so more clearly labeling those variables or testing outputs?)

srtfisher committed:
Apply suggestions from code review

Co-authored-by: Benjamin Bolton <7146063+benpbolton@users.noreply.github.com>

srtfisher committed:
Renaming to include VIP

srtfisher committed:
Code review feedback

srtfisher committed:
Code review feedback contd and add testing for core test suite

srtfisher committed:
Include core test suite in name

srtfisher committed:
Fixing name

srtfisher committed:
Combine actions

srtfisher committed:
Remove set -e

srtfisher committed:
Fix mysql shell logic

srtfisher committed:
Fixing more shellcheck errors

srtfisher committed:
Remove set +e and add debugging

srtfisher committed:
Ignore some errors and fix more

srtfisher committed:
Check if the database was created

srtfisher committed:
Use the root password

srtfisher committed:
Ignore error

srtfisher committed:
Merge remote-tracking branch 'origin/action-shell' into vip-mu-plugins

srtfisher committed:
Switch to localhost

srtfisher committed:
Fixing database check step

srtfisher committed:
Make the clone quiet

srtfisher committed:
Disable fast-fail so we can see all versions

srtfisher committed:
Try 127.0.0.1

srtfisher commented:
@benpbolton would love another pass at this one. I went through and addressed all the feedback you had, removed any `set +e` that I could, and also made it validate `shellcheck` to make it a bit more stable.

While I was here, I also added testing for the core test suite installation as well as the database creation step. I think this PR has made this script quite solid. 😄

benpbolton commented on install-wp-tests.sh:
```diff
@@ -1,52 +1,159 @@
+# Install WordPress Tests for Development
+#
+# Primarily geared for development with Mantle framerwork/Mantle Testkit but also
+# supported installing the core WordPress test suite as well. Mirrors the
```
```suggestion
# supports installing the core WordPress test suite as well. Mirrors the
```

benpbolton commented on install-wp-tests.sh:
```diff
@@ -1,52 +1,159 @@
 	fi
 
-	if [ $(which curl) ]; then
+	if [ "$(which curl)" ]; then
```
```suggestion
	if command -v curl >/dev/null 2>&1; then
```

command -v is the more robust test...

benpbolton commented on install-wp-tests.sh:
```diff
@@ -1,52 +1,159 @@
+	if [ "$(which curl)" ]; then
 		curl -f -s "$1" > "$2"
-	elif [ $(which wget) ]; then
+	elif [ "$(which wget)" ]; then
```
```suggestion
	elif command -v wget >/dev/null 2>&1; then
```

benpbolton commented on install-wp-tests.sh:
```diff
@@ -1,52 +1,159 @@
-	elif [ $(which wget) ]; then
+	elif [ "$(which wget)" ]; then
 		wget -nv -O "$2" "$1"
 	fi
```
Should we add an error exit here (an `else` in the event that neither `curl` nor `wget` are available?

benpbolton commented on .github/workflows/test-script.yml:
```diff
@@ -11,28 +11,33 @@ name: Test Script
+          export INSTALL_WP_TEST_DEBUG=1
+          export WP_INSTALL_CORE_TEST_SUITE=${{ matrix.install-core-tests }}
+
+          bash install-wp-tests.sh wordpress_unit_tests root root localhost "$WP_VERSION" false true true
```
As a human, I wouldn't mind a comment above this line describing the arguments, even if it's just
```
install-wp-tests.sh <db-name> <db-user> <db-pass> <db-host> <wp-version> <skip-database-creation> <install-mu-plugins> <install-memcached>
```

benpbolton reviewed (commented):
Should we add an `else` condition where neither curl nor wget are in place?

srtfisher committed:
Apply suggestions from code review

Co-authored-by: Benjamin Bolton <7146063+benpbolton@users.noreply.github.com>

srtfisher committed:
Last code review feedback

srtfisher committed:
Remove empty lines
